### PR TITLE
Ruby on Rails capstone project - Budget app

### PR DESCRIPTION
**Pull request by lincoln1883**
For this project i have completed all requirements.
- [x] Design including:
  - Colors.
  - Typography: font face, size and weight.
  - Layout: composition and space between elements.
- [x] Interactions
 - [x] Splash screen
  - [x] A simple page with the name of your app (yes, you need to choose one), and links to the 
  - [x] sign up and log in pages.
  - [x] The user should be able to register in the app with full name, email and password (all mandatory).
  - [x] The user can log into the app using email and password.
  - [x] If the user is not logged in, they can't access pages that require the user to be logged in (all the pages described below).
- [x] Home page (categories page)
  - [x] When the user logs in, they are presented with the categories page.
  - [x] For each category, the user can see their name, icon and the total amount of all the transactions that belongs to that category.
  - [x] When the user clicks (or taps) on a category item, the application navigates to the transactions page for that category.
  - [x] There is a button "add a new category" at the bottom that brings the user to the page to create a new category.
- [x] Transactions page
  - [x] For a given category, the list of transactions is presented, ordered by the most recent.
  - [x] At the top of the page the user could see the total amount for the category (sum of all of the amounts of the transactions in that category).
  - [x] There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
  - [x] When the user clicks on the "Back" button (<), the user navigates to the home page.
- [x] "Add a new category" page
  - [x] The user fills out a form to create a new category, indicating their name and icon (both mandatory).
  - [x] The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
  - [x] When the user clicks on the "Back" button (<), the user navigates to the home page.
- [x] "Add a new transaction" page
  - [x] The user fills out a form to create a new transaction with:
    - [x] name (mandatory)
    - [x] amount (mandatory)
    - [x] categories (mandatory at least one)
 - [x] The user click (or taps) the "Save" button to create the new transaction, and is taken to the transactions page for that category.
 - [x] When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.
- [x] Testing requirements
  - [x] Create unit and integration tests for all the most important components of your RoR application.
- [x] Technical requirements
   - [x] You should use Postgres as your database.
   - [x] You should use devise for authentication.
   - [x] You should validate all user input to make sure that anyone with bad intentions cannot compromise your app.
   - [x] You can use a view template engine of your choice (.erb, .slim, .haml).
   - [x] The project should be deployed and accessible online.
   - [x] Your database schema should reflect the following structure:

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,24 +1,164 @@
...
+- [ ] Integration specs for Views and fixing n+1 problems.
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+<!-- LIVE DEMO -->
+
+### Live Demo <a name = "live-demo"></a>
+[Piggy Bank](https://piggycoin.onrender.com)
+
+### Video Presentation <a name = "video-Presentation"></a>
+[Presentation](https://www.loom.com/share/cc953017d78c41049819a9b6e8b02c44?sid=71dcf0c0-cf5a-45b5-aecb-850581679ddd)
```
Comment (omar-muhamad) [inline README.md]: - [ ] As per this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/capstone/rails_capstone.md#:~:text=the%20project%20features.-,you%20should%20also%20highlight%20some%20interesting%20piece%20of%20code%20or%20something%20you%20built%20that%20you%20are%20very%20proud%20of.,-For%20recording%20the), Please highlight some interesting piece of code or something you built that you are very proud of.

File: app/views/categories/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+              <p class="text-sm text-gray-700 uppercase tracking-wide font-semibold mt-2">
+                <%= time_ago_in_words(category.created_at) %> ago
+              </p>
+            </div>
+            <p>
+          </div>
+        </div>
+      <% end %>
+      <%= button_to 'Delete', category, method: :delete, class: 'mr-[45%] ml-[45%]'%>
+    <% end %>
```
Comment (omar-muhamad) [inline app/views/categories/index.html.erb]: - [ ] As per this [requirement](https://github.com/microverseinc/curriculum-rails/blob/main/capstone/rails_capstone.md#:~:text=size%20and%20weight.-,Layout%3A%20composition%20and%20space%20between%20elements.,-NOTE%3A%20In%20these), Please add the correct widths and margin values for the category card to match the given design.

| Given Design  | Your Design |
| ------------- | ------------- |
| image  | image  |

File: app/models/transaction.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class Transaction < ApplicationRecord
+  validates :name, presence: true, length: { minimum: 3 }
+  validates :category_id, presence: true
+  validates :amount, presence: true, numericality: { greater_than: 1 }
+  validates :user_id, presence: true
+
+  belongs_to :category
+  belongs_to :user
```
Comment (omar-muhamad) [inline app/models/transaction.rb]: - [ ] As per this [note](https://github.com/microverseinc/curriculum-rails/blob/main/capstone/rails_capstone.md#:~:text=for%20your%20project%20(-,please%20note%3A%20%22Transaction%22%20is%20a%20name%20already%20used%20by%20ActiveRecord%2C%20so%20using%20it%20as%20a%20name%20for%20your%20model%20and%20table%20will%20result%20in%20an%20error).,-Project%20documentation), "Transaction" is a name already used by ActiveRecord, so using it as a name for your model and table will result in an error, Please change that

**Review by omar-muhamad — CHANGES_REQUESTED** [review]
# Changes Required ♻️
### Hi @lincoln1883,

### Nice job, you are almost there.

cat-driving-serious

## (Highlights) Good Points: :+1:
- Your work is well documented. ✔
- The correct workflow is used.  ✔
- All linters checks are ok. ✔

## (Changes Required) Aspects to improve: ♻️

- [ ] Please check the inline comments below ⬇️.

## [Optional] suggestions:

- Nothing to mention.

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

**Cheers and Happy coding!👏👏👏**

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: app/views/categories/index.html.erb (hunk 2 above)
Comment (lincoln1883) [reply app/views/categories/index.html.erb]: Hello Omar thank you for your recommendation I understand the point you are trying to make however there are no cellphones narrower than 320 px, I'm not sure what size device you are testing with. please check my screen record also I've viewed the website on my cell phone and there no such issues if you wish I could make a video as well, next time to be more transparent with your findings show the entire screen. 


https://github.com/lincoln1883/piggy-banker/assets/84746035/917320b5-a0e8-46d7-83f4-4c0a8dd25c26


Screenshot from 2023-09-18 16-11-50
Screenshot from 2023-09-18 16-11-07

**Review by Munsa1 — APPROVED** [review]
Hi @lincoln1883, 
congrats 2

In my opinion, your project is ready for the presentation!  There is nothing else to say other than... it's time to merge it :shipit: 


Good luck! 🎉 


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**Pull request merged**